### PR TITLE
fix(xmlupload): print reason of error if a resource cannot be created, remodel BaseError (DEV-1977)

### PR DESCRIPTION
**Pull request by jnussbaum**
resolves DEV-1977

### CODE REVIEW

**Comment by jnussbaum** [comment]
@BalduinLandolt @irinaschubert For more context, see the Linear ticket. Generally, I plan to write more informative Linear tickets in the future. That means for you that you can read the ticket before reviewing, and then you know the rationale.

**Review by irinaschubert — DISMISSED** [review]
I like the intention of this PR very much, if the output is more informative this is certainly a nice support for the user! But I agree with all points mentioned by Balduin and have nothing more to add :)

File: src/dsp_tools/models/connection.py [hunk 1]
```diff
@@ -7,6 +7,25 @@
...
+    """
+    if response.status_code != 200:
+        raise BaseError(
+            message="DSP-ERROR: status code=" + str(response.status_code) + "\nMessage:" + response.text,
+            status_code_of_api_response=response.status_code,
+            json_content_of_api_response=response.text,
+            reason_for_failure_from_api_response=response.reason,
+            api_route=response.url
+        )
+
```
Comment (jnussbaum) [inline src/dsp_tools/models/connection.py:28]: This code block is not new, I just moved and renamed `on_api_error()`

File: src/dsp_tools/models/connection.py [hunk 2]
```diff
@@ -119,22 +126,28 @@ def post(self, path: str, jsondata: Optional[str] = None):
         if jsondata is None:
             if self._token is not None:
                 headers = {'Authorization': 'Bearer ' + self._token}
```
Comment (jnussbaum) [inline src/dsp_tools/models/connection.py:128]: The rest of this file, from here on, contains only refactorings: I wanted to rename the response of each request from `req` to `response`. But by doing so, I had to redo the entire indentation.

File: src/dsp_tools/models/sipi.py [hunk 3]
```diff
@@ -3,25 +3,7 @@
...
-    Returns:
-        Knora Error that is being raised
-    """
-
-    if res.status_code != 200:
-        raise BaseError("SIPI-ERROR: status code=" + str(res.status_code) + "\nMessage:" + res.text)
-
-    if 'error' in res:
-        raise BaseError("SIPI-ERROR: API error: " + res.error)
+from dsp_tools.models.connection import check_for_api_error
```
Comment (jnussbaum) [inline src/dsp_tools/models/sipi.py:6]: I thought it is not necessary that the `Sipi` class contains a duplicate of `check_for_api_error()`. It is the same logic, just that it's a SIPI error, not an API error.

File: src/dsp_tools/utils/shared.py [hunk 4]
```diff
@@ -75,7 +75,7 @@ def try_network_action(action: Callable[..., Any]) -> Any:
             time.sleep(2 ** i)
             continue
         except BaseError as err:
-            if regex.search(r"try again later", err.message) or regex.search(r"status code=5\d\d", err.message):
+            if regex.search(r"try again later", err.message) or regex.search(r"5\d\d", str(err.status_code_of_api_response)):
```
Comment (jnussbaum) [inline src/dsp_tools/utils/shared.py:78]: perhaps not the most elegant way. But if I wanted to analyze `err.status_code_of_api_response` as integer, 
 - I first have to check if it is not None, 
 - then check that it's > 500
 - then check that it's < 600

**Comment by jnussbaum** [comment]
@irinaschubert @BalduinLandolt As suggested by you, I have remodeled `BaseError`. I think it was worth the effort, I very much like the idea of storing fine-grained information in the class `BaseError`.

**Review by BalduinLandolt — APPROVED** [review]
Definitely a step in the right direction. I added two minor comments, but over all looking good.

**Pull request merged**